Backend/FcTypeChecker.hs

```diff
@@ -354,22 +364,34 @@ matchClause dc (u:us) qs def = do
 matchClause _   []    _  _   = panic "matchClause: empty variables"
 
 -- | Match equations according to (new) or rule
-matchOr :: [FcTmVar] -> [PmEqn] -> FcTerm 'Fc -> FcM (FcTerm 'Fc)
-matchOr us ((((FcOrPat p1 p2):ps), rhs):qs) def = match us (((p1:ps), rhs) : ((p2:ps), rhs) : qs) def
-matchOr _ _ _ = panic ("matchOr: called on a non or-pattern equation")
+matchOr :: [FcTmVar] -> ([PmEqn], PmEqn, [PmEqn]) -> FcTerm 'Fc -> FcM (FcTerm 'Fc)
+matchOr (u:us) (preq, (((FcOrPat p1 p2):ps), rhs), postq) def = do
+  exp_ty      <- lookupTmVarM u
+  tmVarTys    <- extractTmVarTys exp_ty (FcOrPat p1 p2)
+  (tm, tm_ty) <- extendCtxTmZipM tmVarTys (match us [(ps, rhs)] def >>= tcTerm)
+  f           <- foldrM abstractTmVarTy tm tmVarTys
+  let f_ty    = foldr (\(_,ty1) ty2 -> mkFcArrowTy ty1 ty2) tm_ty tmVarTys
+  y           <- makeVar
+  let y_app   = foldl applyTmVarTy (FcTmVar y) tmVarTys
+  let g_app   = FcGuarded [] y_app
+  matched     <- extendCtxTmM y f_ty (match (u:us) (preq ++ [(p1:ps, [g_app]), (p2:ps, [g_app])] ++ postq) def)
+  return (FcTmLet y f_ty f matched)
+matchOr [] _ _ = panic ("matchOr: empty variable")
+matchOr _ _ _ = panic ("matchOr: no or pattern in equation")
```

Comment:
Best if you explain using comments what the approach is here; it's quite difficult to understand what this function does exactly.

Reply:
I wrote some comments, do they help? 😄